src/polyasm/polyasm.ts

```diff
@@ -0,0 +1,884 @@
+			this.consume(1);
+		}
+	}
+
+	public getInstructionSize(): number {
+		try {
+			const instructionTokens = this.getInstructionTokens();
+			const mnemonicToken = instructionTokens[0] as ScalarToken;
+			const operandTokens = instructionTokens.slice(1) as OperatorStackToken[];
+
+			const sizeInfo = this.cpuHandler.resolveAddressingMode(mnemonicToken.value, operandTokens, (exprTokens) =>
+				this.expressionEvaluator.evaluateAsNumber(exprTokens, {
+					pc: this.currentPC,
+					macroArgs: this.tokenStreamStack[this.tokenStreamStack.length - 1].macroArgs,
+					currentGlobalLabel: this.lastGlobalLabel, // Added for instruction size evaluation
+					options: this.options,
+				}),
+			);
+			return sizeInfo.bytes;
+		} catch (_e) {
+			return this.cpuHandler.cpuType === "ARM_RISC" ? 4 : 3; // Robust default based on CPU type
+		}
+	}
+
+	private extractExpressionArrayTokens(tokens: Token[]) {
+		const result: Token[][] = [];
+		let current: Token[] = [];
+		let parenDepth = 0;
+
+		if (tokens.length < 2) return result;
+		if (tokens[0].value !== "[" || tokens[tokens.length - 1].value !== "]") return result;
+
+		// for (const token of tokens) {
+		for (let idx = 1; idx < tokens.length - 1; idx++) {
+			const token = tokens[idx];
+			if (token.type === "OPERATOR" && token.value === "(") parenDepth++;
+			if (token.type === "OPERATOR" && token.value === ")") parenDepth--;
+			if (token.type === "COMMA" && parenDepth === 0) {
+				result.push(current);
+				current = [];
+			} else {
+				current.push(token);
+			}
+		}
+		if (current.length > 0) result.push(current);
+
+		return result;
+	}
+
+	private substituteTokens(tokens: Token[], macroArgs: Map<string, Token[]>): Token[] {
+		const result: Token[] = [];
+		for (let i = 0; i < tokens.length; i++) {
+			const token = tokens[i];
+
+			if (token.type !== "IDENTIFIER" || !macroArgs.has(token.value)) {
+				result.push(token);
+				continue;
+			}
+
+			// Token is a macro argument.
+
+			// Check for array access like `parms[0]`
+			if (i + 1 < tokens.length && tokens[i + 1].value === "[") {
+				let j = i + 2;
+				let parenDepth = 0;
+				const indexTokens = [];
+
+				// Find closing ']' and gather index tokens
+				while (j < tokens.length) {
+					if (tokens[j].value === "[") parenDepth++;
+					else if (tokens[j].value === "]") {
+						if (parenDepth === 0) break;
+						parenDepth--;
+					}
+					indexTokens.push(tokens[j]);
+					j++;
+				}
+
+				// If we found a complete `[...]` expression
+				if (j < tokens.length && tokens[j].value === "]") {
+					const indexValue = this.expressionEvaluator.evaluateAsNumber(indexTokens, {
+						pc: this.currentPC,
+						macroArgs: macroArgs, // Pass current macro args for evaluation context
+						assembler: this,
+						currentGlobalLabel: this.lastGlobalLabel,
+						options: this.options,
+					});
+
+					const argTokens = macroArgs.get(token.value) ?? [];
+					const expressions = this.extractExpressionArrayTokens(argTokens);
```

Comment:
here, for the rest parameter, we're deconstructing what we built in the expnadMacro. Surely, there's a better way. Like having directly an array of expressions.
What happens for array argument, like myMacro([1,2,3]) ?